pykern/fconf.py

```diff
@@ -0,0 +1,517 @@
+# -*- coding: utf-8 -*-
+"""Python and YAML configuration file parser.
+
+FConf reads Python and YAML files and produces a single, merged
+PKDict. FConf is not a replacement for `pykern.pkconfig`. Rather it is
+for complex configuration input files to programs that often require
+programmatic generation. FConf was written for
+RSConf <https://git.radiasoft.org/rsconf>.
+
+The Basic YAML configuration look like this::
+
+    fconf_macros:
+      make_uri(host):
+          "https://${host}"
+
+    some:
+       - item
+    uri: make_uri('sirepo.com')
+
+Macros in the `fconf_macros` are simple replacements.  They cannot have
+loops or other conditions. These are reserved for Python files.
+
+A macro use must be a full text element. It cannot be embedded in
+text, e.g.  ``curl make_uri('sirepo.com')``, that's because macros can
+return data structures. The syntax is pure Python. The arguments have
+to quoted.
+
+More complex macros can be programmed in a Python file::
+
+    import sqrt
+
+    def hypotenuse(self, a, b):
+        return math.sqrt(a ** 2, b ** 2)
+
+Using in YAML looks the same::
+
+    long_side: hypotenuse(3, 4)
+
+Note that the arguments were not quoted here, because they are integers.
+
+The typical usage is::
+
+    fconf.parse_all(some_dir)
+
+where `some_dir` is a `py.path` that contains python and yaml files to parse.
+You can also call `Parser` directly with the a list of files to parse.
+
+As noted, the arguments to a macro are Python. Actually, the entire
+value is a Python expression, and the only requirement is that it
+begin with a Python word, which can be any Python function. For example::
+
+    count_down:
+      - reversed(range(4))
+
+Produces the following Python Structure::
+
+    {'count_down': [3, 2, 1, 0]}
+
+This allows for a lot of possibilities within the YAML alone.
+
+The Python and YAML macros can call each other. Inside a Python file,
+just use self to call something::
+
+    def count_up(self, top):
+        return reversed(self.count_down(top))
+
+Note that the above count_down example used a special feature of
+result merging. When a Macro returns a list (generator, tuple, etc.)
+in a list context, it will merge with the list. Strings are used in
+place. For example::
+
+    mixed:
+      - range(3)
+      - make_uri('sirepo.com')
+
+Produces::
+
+    {'mixed': [0, 1, 2, 'https://sirepo.com']}
+
+The same is true of dicts. For example::
+
+    fconf_macros:
+      base(num):
+        v${num}.radia.run:
+          description: VM ${num}
+
+    all_hosts:
+      base(3):
```

Comment:
I didn't know this was valid yaml. Funky.

This example was my "aha" moment. This would be such a pain with jinja due to indenting.

Reply:
It's actually a very interesting construct, because:

```yml
x:
```

generates `{"x": None}` which is used to enforce that macro uses as keys that result in dicts have None as values. This will allow in some future version (if necessary):

```yml
x:
   y: 3
```

where the `{"y": 3}` would be available to the macro implementation to embed (YAML macros) or to evaluate (Python macros). The latter case would allow control structures like real macros.